### PR TITLE
LA-135: add network state connection observer

### PR DESCRIPTION
**Pull request by manhduc-nguyen**
Screen Shot 2021-01-25 at 17 25 35

### CODE REVIEW

File: lib/presentation/widget/home/home_viewmodel.dart [hunk 1]
```diff
@@ -69,6 +74,19 @@ class HomeViewModel extends BaseViewModel {
     });
   }
 
+  void _registerNetworkConnectivityState() async {
+    store.dispatch(_setNetworkStateAction(await Connectivity().checkConnectivity()));
+    _connectivityStreamSubscription = Connectivity().onConnectivityChanged.listen((ConnectivityResult result) {
```
Comment (hoangdat) [inline lib/presentation/widget/home/home_viewmodel.dart:79]: Please provide `Connectivity` as a singleton.

File: lib/presentation/redux/states/network_connectivity_state.dart [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+class NetworkConnectivityState extends LinShareState {
+  final ConnectivityResult connectivityResult;
+
+  NetworkConnectivityState(Either<Failure, Success> viewState, this.connectivityResult) : super(viewState);
+
+  factory NetworkConnectivityState.initial() {
+    return NetworkConnectivityState(Right(IdleState()), null);
+  }
+
+  LinShareState setNetworkState(ConnectivityResult connectivityResult) {
```
Comment (imGok) [inline lib/presentation/redux/states/network_connectivity_state.dart:50]: Methods in state should return ``NetworkConnectivityState`` imo

File: lib/presentation/redux/states/network_connectivity_state.dart [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+import 'package:connectivity/connectivity.dart';
+import 'package:dartz/dartz.dart';
+import 'package:domain/src/state/failure.dart';
+import 'package:domain/src/state/success.dart';
+import 'package:flutter/foundation.dart';
+import 'package:linshare_flutter_app/presentation/redux/states/linshare_state.dart';
+
+@immutable
+class NetworkConnectivityState extends LinShareState {
+  final ConnectivityResult connectivityResult;
```
Comment (imGok) [inline lib/presentation/redux/states/network_connectivity_state.dart:42]: I don't think creating a state just for that is good why not put it in UI state ?

File: lib/presentation/widget/home/home_viewmodel.dart [hunk 4]
```diff
@@ -69,6 +76,19 @@ class HomeViewModel extends BaseViewModel {
...
+  void _registerNetworkConnectivityState() async {
+    store.dispatch(_setNetworkStateAction(await _connectivity.checkConnectivity()));
+    _connectivityStreamSubscription = _connectivity.onConnectivityChanged.listen((ConnectivityResult result) {
+      store.dispatch(_setNetworkStateAction(result));
+    });
+  }
+
+  ThunkAction<AppState> _setNetworkStateAction(ConnectivityResult connectivityResult) {
+    return (Store<AppState> store) async {
+      store.dispatch(SetNetworkConnectivityStateAction(connectivityResult));
```
Comment (imGok) [inline lib/presentation/widget/home/home_viewmodel.dart:88]: You can directly send this action you don't need a ThunkAction wdyt ?

File: lib/presentation/redux/states/network_connectivity_state.dart (hunk 3 above)
Comment (manhduc-nguyen) [reply lib/presentation/redux/states/network_connectivity_state.dart:43]: Oh no, I think its not really related to `UIState`. We should separate it.

File: lib/presentation/redux/states/network_connectivity_state.dart (hunk 2 above)
Comment (manhduc-nguyen) [reply lib/presentation/redux/states/network_connectivity_state.dart]: you are right.

**Comment by manhduc-nguyen** [comment]
@imGok @hoangdat pls check my fix up commit. I also added unit test

**Review by imGok — APPROVED** [review]

**Review by hoangdat — APPROVED** [review]

**Pull request merged**